_posts/2021-01-08-consumer-tuning.md

```diff
@@ -0,0 +1,280 @@
+
+* **`max.partition.fetch.bytes`**
+Sets a maximum limit in bytes on how much data is returned for each partition, which must always be larger than the number of bytes set in the broker or topic configuration for `max.message.bytes`.
+
+By increasing the values of these two properties, and allowing more data in each request, latency is improved as there are fewer fetch requests.
+
+```properties
+# ...
+fetch.max.bytes=52428800
+max.partition.fetch.bytes=1048576
+# ...
+```
+
+> Be careful with memory usage. The maximum amount of memory a client can consume is calculated approximately as:
+>
+> `_NUMBER-OF-BROKERS_ * fetch.max.bytes and _NUMBER-OF-PARTITIONS_ * max.partition.fetch.bytes`
+
+
+## Avoiding data loss or duplication when committing offsets
+
+For applications that require durable message delivery, you can increase the level of control over consumers when committing offsets to minimize the risk of data being lost or duplicated.
+
+Kafka's _auto-commit mechanism_ is pretty convenient (and sometimes suitable). When enabled, consumers commit the offsets of messages automatically. But convenience, as always, has a price. By allowing your consumer to commit offsets, you are introducing a risk of data loss and duplication. Why? Quite easily as it happens. You see, the auto-commit can _only_ avoid data loss if all messages are processed before the next poll of the broker or the consumer closes. Using the auto-commit, the broker is polled at 5000ms intervals. In that time, the inherent constraints of auto-committing might lead to:
+
+* _Data loss_ when the consumer has fetched and transformed a number of messages, but the system crashes with processed messages still in the consumer buffer
+* _Data duplication_ on another consumer instance, after rebalancing, when the system crashes whilst processing the messages, but before performing the auto-commit
+
+If this potential situation leaves you slightly concerned, what can you do about it? First of all, you can simply turn off auto-committing by setting `enable.auto.commit` to `false`. You are now in the driving seat. What else? Alternatively, you can use the `auto.commit.interval.ms` property to decrease those worrying intervals between commits.
+
+```properties
+# ...
+enable.auto.commit=false
+auto.commit.interval.ms=1000
+# ...
+```
+
+A more robust course of action, away from consumer configuration, is to set up your consumer client application to commit offsets after _all_ processing has been performed and messages have been consumed. To do this, you can introduce calls to the Kafka `commitSync` and `commitAsync` APIs, which commit specified offsets for topics and partitions to Kafka.
+
+* **`commitSync`**
+The `commitSync` API will not poll for new messages until the last offset in a message batch is committed. This can affect throughput and latency, but you can configure the consumer to commit less frequently.
```

Comment:
I don't think that this API is related to poll or it avoids you to poll for new messages while committing.
I mean, if I write an application where in two different threads I do commitSync and poll I guess I can do that. Of course, it's just an example, something I would not do but just to explain that the sentence sounds misleading to me. Maybe it's still related to the fact that we should pay more attention when the "poll" verb is used because it has a specific meaning in Kafka. @tombentley wdyt?

Reply:
Yeah, this sentence is confusing. `commitSync` doesn't call `poll()`, but rather commits the offsets of all the messages which were returned by the last `poll()`. " you can configure the consumer to commit less frequently" is also a bit ambiguous. I assume the context here is a consumer application where we're doing a "manual commit", in which case it's not really configuration, but rather how that application is coded. I.e. is it committing after every poll, or less frequently. And how many messages are returned by each poll? 

We also don't talk about an important alternative to committing offsets in Kafka when interacting with some other system which supports ACID transactions. In that case you can get better processing guarantees by storing the offsets in that system. I'm not saying we should talk about this in detail, because the whole topic can mushroom into its own blog post, merely highlighting that it might be worth mentioning in a sentence.

Reply:
Thanks. Trying to to chop the concepts down so much didn't help here. Hopefully, this get's closer to what we want to say.

> * **`commitSync`**
> The `commitSync` API commits the offsets of all messages returned from polling. Generally, you call the API when you are finished processing all the messages in a batch and don't poll for new messages until the last offset in the batch is committed. This approach can can affect throughput and latency, as can the number of messages returned when polling, so you can set up your application to commit less fequently. 

And I've added a note on ACID.

> You can set up your Kafka solution to interact with databases that provide ACID
> (atomicity, consistency, isolation, and durability) reliability
> guarantees to store consumer offsets.